blmath/geometry/primitives/plane.py

```diff
@@ -221,157 +260,104 @@ def mesh_xsection(self, m, neighborhood=None):
+        def edge_from_face(f):
+            face_verts = [
+                [m.v[f[0]], m.v[f[1]]],
+                [m.v[f[1]], m.v[f[2]]],
+                [m.v[f[2]], m.v[f[0]]],
+            ]
+            e = [self.line_segment_xsection(a, b) for a, b in face_verts]
+            e = [x for x in e if x is not None]
+            return e
+        edges = np.vstack([np.hstack(edge_from_face(f)) for f in fs])
+
+        # 3: Find the set of unique vertices in `edges`
+        v1s, v2s = np.hsplit(edges, 2)
+        verts = edges.reshape((-1, 3))
+        verts = np.vstack(sorted(verts, key=operator.itemgetter(0, 1, 2)))
+        eps = 1e-15 # the floating point calculation of the intersection locations is not _quite_ exact
+        verts = verts[list(np.sqrt(np.sum(np.diff(verts, axis=0) ** 2, axis=1)) > eps) + [True]]
+        # the True at the end there is because np.diff returns pairwise differences; one less element than the original array
+
+        # 4: Build the edge adjacency matrix
+        E = sp.dok_matrix((verts.shape[0], verts.shape[0]), dtype=np.bool)
+        def indexof(v, in_this):
+            return np.nonzero(np.all(np.abs(in_this - v) < eps, axis=1))[0]
+        for ii, v in enumerate(verts):
+            for other_v in list(v2s[indexof(v, v1s)]) + list(v1s[indexof(v, v2s)]):
+                neighbors = indexof(other_v, verts)
+                E[ii, neighbors] = True
+                E[neighbors, ii] = True
+
+        def eulerPath(E):
+            # Based on code from Przemek Drochomirecki, Krakow, 5 Nov 2006
+            # http://code.activestate.com/recipes/498243-finding-eulerian-path-in-undirected-graph/
+            # Under PSF License
+            # NB: MUTATES graph
+            if len(E.nonzero()[0]) == 0:
+                return None
+            # counting the number of vertices with odd degree
+            odd = list(np.nonzero(np.bitwise_and(np.sum(E, axis=0), 1))[0])
+            odd.append(np.nonzero(E)[0][0])
+            if len(odd) > 3:
+                return None
+            stack = [odd[0]]
+            path = []
+            # main algorithm
+            while stack:
+                v = stack[-1]
+                nonzero = np.nonzero(E)
+                nbrs = nonzero[1][nonzero[0] == v]
+                if len(nbrs) > 0:
+                    u = nbrs[0]
+                    stack.append(u)
+                    # deleting edge u-v
+                    E[u, v] = False
+                    E[v, u] = False
+                else:
+                    path.append(stack.pop())
+            return path
+
+        # 5: Find the paths for each component
+        components = []
+        while len(E.nonzero()[0]) > 0:
+            # This works because eulerPath mutates the graph as it goes
+            path = eulerPath(E)
+            if path is None:
+                raise ValueError("mesh slice has too many odd degree edges; can't find a path along the edge")
```

Comment:
Will need a different way to break out of the loop too I think -- when I take out the `raise` this `while` loop just keeps running